### PR TITLE
Use protocol

### PR DESCRIPTION
**Pull request by duanxianbo**

### CODE REVIEW

File: apps/github/app_api.go [hunk 1]
```diff
@@ -461,6 +462,27 @@ func (a AppAPI) webListTeamMembers(writer http.ResponseWriter, request *http.Req
 		return
 	}
 
+	userIDs := collect.Map(listTeamMembersRes.TeamMembers, func(member *pbmessage.TeamMember, _ int) uint64 {
+		return member.UserId
+	})
+
```
Comment (naturetime10) [inline apps/github/app_api.go]:
```suggestion
```

File: apps/github/app_api.go [hunk 2]
```diff
@@ -461,6 +462,27 @@ func (a AppAPI) webListTeamMembers(writer http.ResponseWriter, request *http.Req
...
+	listUsersReq := &pbteamy.ListUsersRequest{
+		UserIds: userIDs,
+	}
+
```
Comment (naturetime10) [inline apps/github/app_api.go]:
```suggestion
```

File: apps/github/app_api.go [hunk 3]
```diff
@@ -461,6 +462,27 @@ func (a AppAPI) webListTeamMembers(writer http.ResponseWriter, request *http.Req
 		return
 	}
 
+	userIDs := collect.Map(listTeamMembersRes.TeamMembers, func(member *pbmessage.TeamMember, _ int) uint64 {
```
Comment (naturetime10) [inline apps/github/app_api.go:465]: Can we fetch users directly?

File: core/api/task_rpc.go [hunk 4]
```diff
@@ -100,23 +106,39 @@ func (t TaskRPC) CreateTask(ct context.Context, req *pbteamy.CreateTaskRequest)
 	return &pbteamy.CreateTaskResponse{TaskId: task.ID}, nil
 }
 
-func (t TaskRPC) UpdateTask(ct context.Context, req *pbteamy.UpdateTaskRequest) (*emptypb.Empty, error) {
+func (t TaskRPC) UpdateTask(ct context.Context, req *pbteamy.UpdateTaskRequest) (*pbteamy.UpdateTaskResponse, error) {
 	input := service.UpdateTaskInput{
-		Goal:         req.Goal,
+		Goal:         *req.Goal,
```
Comment (naturetime10) [inline core/api/task_rpc.go]: Is this `*` needed? Will this cause `nil` pointer error?

File: core/api/user_rpc.go [hunk 5]
```diff
@@ -0,0 +1,56 @@
...
+func (u UserRPC) Start(runner *runner.ServiceRunner) *errs.Error {
+	runner.WithGRPCServer(func(server *grpc.Server) {
+		pbteamy.RegisterUserServiceServer(server, u)
+	})
+	return nil
+}
+
+func (u UserRPC) ListUsers(ct context.Context, listUsersRequest *pbteamy.ListUsersRequest) (*pbteamy.ListUsersResponse, error) {
+	userIDs := make([]uint64, 0, len(listUsersRequest.UserIds))
+
```
Comment (naturetime10) [inline core/api/user_rpc.go]:
```suggestion
```

File: go.mod [hunk 6]
```diff
@@ -17,7 +17,8 @@ require (
 	gopkg.in/yaml.v3 v3.0.1
 )
 
-//replace github.com/teamyapp/cloud => ../cloud
+// replace github.com/teamyapp/cloud => ../cloud
+replace github.com/teamyapp/protocol/pb/pbgo => ../protocol/pb/pbgo
```
Comment (naturetime10) [inline go.mod]: Remember to check in protocol first.

**Review by naturetime10 — CHANGES_REQUESTED** [review]

File: core/api/task_rpc.go (hunk 4 above)
Comment (duanxianbo) [reply core/api/task_rpc.go]: make every update field nullable

File: core/service/task.go [hunk 7]
```diff
@@ -562,10 +562,14 @@ func (t Task) UpdateTask(ct context.Context, taskID uint64, input UpdateTaskInpu
 
 			oldEffort := task.Effort
 			oldOwnerID := task.OwnerUserID
-			task.Goal = input.Goal
+			if input.Goal != nil {
+				task.Goal = *input.Goal
+			}
```
Comment (naturetime10) [inline core/service/task.go:567]:
```suggestion
			}
			
```

File: core/service/task.go [hunk 8]
```diff
@@ -562,10 +562,14 @@ func (t Task) UpdateTask(ct context.Context, taskID uint64, input UpdateTaskInpu
...
 			task.Context = input.Context
 			task.OwnerUserID = input.OwnerUserID
-			task.OwningTeamID = input.OwningTeamID
+			if input.OwningTeamID != nil {
+				task.OwningTeamID = *input.OwningTeamID
+			}
```
Comment (naturetime10) [inline core/service/task.go:573]:
```suggestion
			}
			
```

File: core/api/task_rpc.go [hunk 9]
```diff
@@ -189,9 +211,11 @@ func (t TaskRPC) RemoveAwaitForTask(ct context.Context, req *pbteamy.RemoveAwait
 	return &emptypb.Empty{}, nil
 }
 
-func NewTaskRPC(logger telemetry.Logger, taskService service.Task) TaskRPC {
+func NewTaskRPC(logger telemetry.Logger, taskService service.Task, taskLinkService service.TaskLink, attachmentService *service.Attachment) TaskRPC {
```
Comment (naturetime10) [inline core/api/task_rpc.go]:
```suggestion
func NewTaskRPC(
    logger telemetry.Logger, 
    taskService service.Task, 
    taskLinkService service.TaskLink, 
    attachmentService *service.Attachment,
) TaskRPC {
```

**Review by naturetime10 — APPROVED** [review]

File: core/api/gql/schema.graphql [hunk 10]
```diff
@@ -750,10 +750,10 @@ input CreateTaskLinkInput {
 }
 
 input UpdateTaskInput {
-	goal: String!
+	goal: String
 	context: String
 	ownerUserId: ID
-	owningTeamId: ID!
+	owningTeamId: ID
```
Comment (naturetime10) [inline core/api/gql/schema.graphql:756]: This is changing the semantics for update input. Is this a pattern change moving forwards?

**Review by naturetime10 — APPROVED** [review]

File: core/api/gql/schema.graphql (hunk 10 above)
Comment (duanxianbo) [reply core/api/gql/schema.graphql:756]: I think so, to generalize this pattern will make our migration easier

**Pull request merged**